libraries/lib-time-and-pitch/FormantShifter.cpp

```diff
@@ -0,0 +1,183 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  FormantShifter.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "FormantShifter.h"
+#include "FormantShifterLoggerInterface.h"
+#include <algorithm>
+#include <fstream>
+
+namespace
+{
+constexpr auto MapToPositiveHalfIndex(int index, int fullSize)
+{
+   const auto inputIsValid = fullSize > 0 && fullSize % 2 == 0;
+   if (!inputIsValid)
+      return 0;
```

Comment:
I didn't know `expected`, that's interesting, thank you. Unfortunately we still C++17, and even our CI toolchain for MacOS doesn't support the full C++17 std lib, e.g. `<filesystem>`. I don't know the exact version numbers, but the oldest MacOS version we release Audacity for cannot deal with the most recent C++ versions.
Also, we are relatively lenient in terms of safety: we document functions with `@pre` conditions, assert these preconditions, but generally don't otherwise deal with error handling. You won't find this pattern often, where the return value is checked for error. Seems to work for us 🤷 On the other hand, unit-testing is encouraged, and I have been lazy doing this time...